ext/ddtrace_profiling_native_extension/collectors_discrete_dynamic_sampler.c

```diff
@@ -0,0 +1,323 @@
+
+  #ifdef DD_DEBUG
+    double allocs_in_60s = sampler->events_per_ns * 1e9 * 60;
+    double samples_in_60s = allocs_in_60s * sampler->sampling_probability;
+    double expected_total_sampling_time_in_60s =
+      samples_in_60s * sampler->sampling_time_ns / 1e9;
+    double real_total_sampling_time_in_60s = sampling_window_time_ns / 1e9 * 60 / (window_time_ns / 1e9);
+
+    fprintf(stderr, "[dds.%s] readjusting...\n", sampler->id);
+    fprintf(stderr, "samples_since_last_readjustment=%ld\n", sampler->samples_since_last_readjustment);
+    fprintf(stderr, "window_time=%ld\n", window_time_ns);
+    fprintf(stderr, "events_per_sec=%f\n", sampler->events_per_ns * 1e9);
+    fprintf(stderr, "sampling_time=%ld\n", sampler->sampling_time_ns);
+    fprintf(stderr, "sampling_window_time=%ld\n", sampling_window_time_ns);
+    fprintf(stderr, "sampling_overshoot_time=%ld\n", sampling_overshoot_time_ns);
+    fprintf(stderr, "working_window_time=%ld\n", working_window_time_ns);
+    fprintf(stderr, "sampling_interval=%zu\n", sampler->sampling_interval);
+    fprintf(stderr, "sampling_probability=%f\n", sampler->sampling_probability);
+    fprintf(stderr, "expected allocs in 60s=%f\n", allocs_in_60s);
+    fprintf(stderr, "expected samples in 60s=%f\n", samples_in_60s);
+    fprintf(stderr, "expected sampling time in 60s=%f (previous real=%f)\n", expected_total_sampling_time_in_60s, real_total_sampling_time_in_60s);
+    fprintf(stderr, "target_overhead_adjustment=%f\n", sampler->target_overhead_adjustment);
+    fprintf(stderr, "expected max overhead in 60s=%f\n", target_overhead / 100.0 * 60);
+    fprintf(stderr, "-------\n");
+  #endif
+
+  sampler->events_since_last_readjustment = 0;
+  sampler->samples_since_last_readjustment = 0;
+  sampler->sampling_time_since_last_readjustment_ns = 0;
+  sampler->last_readjust_time_ns = now;
+}
+
+// ---
+// Below here is boilerplate to expose the above code to Ruby so that we can test it with RSpec as usual.
+
+static VALUE _native_new(VALUE klass);
+static VALUE _native_reset(VALUE self, VALUE target_overhead);
+static VALUE _native_should_sample(VALUE self, VALUE now);
+static VALUE _native_after_sample(VALUE self, VALUE now);
+static VALUE _native_probability(VALUE self);
+
+void collectors_discrete_dynamic_sampler_init(VALUE profiling_module) {
+  VALUE collectors_module = rb_define_module_under(profiling_module, "Collectors");
+  VALUE testing_module = rb_define_module_under(collectors_module, "Testing");
+  VALUE sampler_class = rb_define_class_under(testing_module, "DiscreteDynamicSampler", rb_cObject);
+
+  rb_define_alloc_func(sampler_class, _native_new);
+
+  rb_define_method(sampler_class, "reset", _native_reset, 1);
+  rb_define_method(sampler_class, "should_sample", _native_should_sample, 1);
+  rb_define_method(sampler_class, "after_sample", _native_after_sample, 1);
+  rb_define_method(sampler_class, "probability", _native_probability, 0);
```

Comment:
Minor: I've usually always bound native methods as `_native_foo` on the Ruby side, to make it more clear when we're going from Ruby code to native code. I think it would be nice to do it for this one too? (Although happy to revisit if you find this approach a bit too ugly/heavy-handed)

Reply:
Hmm yeah I wasn't aware the intention was to actually signal the Ruby<>native barrier. Thought it was a kind of differentiator to not clash with the pure ruby methods. And since in this case the object has no backing .rb file, I had dropped them.

The signalling of native barrier intention definitely makes sense so I'll change this!

Reply:
Oh, yeah, in hindsight it wasn't very clear. I'm not a fan of playing the "hey, where the hell is this method defined?" game and hence I went with this to preempt such chases ;)